### PR TITLE
config: Support the Cherami transport

### PR DESCRIPTION
**Pull request by abhinav**
This adds support for the Cherami transport to the configuration system
defined in #747.

The InboundConfig and OutboundConfig structs used by the Cherami
NewInbound and NewOutbound constructors were renamed to InboundOptions
and OutboundOptions to free up the *Config names for use by the
configuration system. This was to keep the naming consistent with the
HTTP (#748), Redis (#749), and TChannel (#750) configuration systems.

### CODE REVIEW

**Comment by breerly** [comment]
@datoug we're taking on getting the Cherami transport to work with the new config paradigm. This will make rpcinit 2.0 "just work" for Cherami.

File: transport/x/cherami/config.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		ts.defaultPeerList = path
+	}
+}
+
+// TransportSpec builds a TransportSpec for the Cherami transport.
+// TransportSpecOptions may be passed to this function to configure the
+// behavior of the TransportSpec.
+//
+// 	configurator.MustRegisterTransport(
+// 		cherami.TransportSpec(cherami.DefaultPeerList("/etc/hosts.json")),
```
Comment (breerly) [inline transport/x/cherami/config.go]: Great docs

File: transport/x/cherami/config.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+)
+
+var errNoPeerList = errors.New(
+	"cannot automatically discover Cherami: " +
+		"no peer list provided: " +
+		"please set the peerList attribute or " +
+		"provide a default peer list using the DefaultPeerList option")
+
+// TransportSpecOption configures the Cherami TransportSpec.
+type TransportSpecOption func(*transportSpec)
```
Comment (breerly) [inline transport/x/cherami/config.go]: I half expected to see an option where you could pass in a Cherami client.

File: transport/x/cherami/config.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/x/config"
+
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+const (
+	_destinationSuffix   = "yarpc_dest"
+	_consumerGroupSuffix = "yarpc_cg"
```
Comment (breerly) [inline transport/x/cherami/config.go]: Honestly I've been wondering if having these not be explicit does more harm than good.

If Cherami created these automatically, then it makes sense.

But because they dont, everyone we've showed has asked "what is the destination group?" and "what is the consumer group?" - I think in this case the default isn't helping us.

@datoug do you agree?

File: transport/x/cherami/config.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+		BuildOnewayOutbound: ts.buildOnewayOutbound,
+	}
+}
+
+// TransportConfig configures the shared Cherami Transport. This is shared
+// between all Cherami outbounds and inbounds of a Dispatcher.
+//
+// All fields in TransportConfig are optional and this section may be skipped
+// entirely for most use cases.
+type TransportConfig struct {
```
Comment (breerly) [inline transport/x/cherami/config.go]: So exciting to see the config entirely encapsulated in the Cherami transport.

File: transport/x/cherami/config.go (hunk 2 above)
Comment (abhinav) [inline transport/x/cherami/config.go]: That was intentionally left out. We can add it later if necessary, or I can add it to this PR if we know we need to support it already.

Comment (datoug) [inline transport/x/cherami/config.go]: +0.5.

**Comment by datoug** [comment]
Generally looks good. Will take another look and stamp once this gets out of WIP

File: transport/x/cherami/config.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+	// 	timeout: 5s
+	Timeout time.Duration `config:"timeout"`
+
+	// Name of the Cherami deployment to which requests should be sent. Some
+	// valid values are, "prod", "staging", "staging2", and "dev".
+	//
+	// 	deploymentStr: dev
+	//
+	// Defaults to "prod".
+	DeploymentStr string `config:"deploymentStr"`
```
Comment (peter-edge) [inline transport/x/cherami/config.go]: I don't like the `Str` suffix, this isn't like anything we do in other areas. It would be nice if `DeploymentType` was an enum, ie `type DeploymentType int` and we could parse this from a string.

**Comment by abhinav** [comment]
@yarpc/golang Ready to merge

File: transport/x/cherami/config.go (hunk 2 above)
Comment (breerly) [inline transport/x/cherami/config.go]: Seems pretty likely that we'll want that, but agreed that we can easily add it whenever.

**Comment by breerly** [comment]
(please raise code coverage above 80% here)

File: transport/x/cherami/config.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	PeerList string `config:"peerList,interpolate"`
+
+	// Timeout for requests to the Cherami service. The default timeout should
+	// suffice for most use cases.
+	//
+	// 	timeout: 5s
+	Timeout time.Duration `config:"timeout"`
+
+	// Name of the Cherami deployment to which requests should be sent. Some
+	// valid values are, "prod", "staging", "staging2", and "dev".
```
Comment (datoug) [inline transport/x/cherami/config.go]: Nit: regarding "Some valid values are"
Since this is the open source repo, we probably reword this to:
"Some examples are"

**Comment by abhinav** [comment]
@yarpc/golang I added a bunch of tests if anyone wants to review them
